File: CSharp/Algorithms/DFS.cs
using System;
using System.Collections.Generic;

public class DFS
{
    public class Graph
    {
        public LinkedList<int>[] adj;
        public int Size;
    }

    public static void dfs(Graph G, int startVert)
    {
        bool[] visited = new bool[G.Size];
        // TODO change this for release. Currently references local Queue implementation otherwise.
        System.Collections.Generic.Stack<int> s = new System.Collections.Generic.Stack<int>();

        visited[startVert] = true;

        s.Push(startVert);

        while(s.Count > 0)
        {
            int v = s.Pop();
            foreach(int adjV in G.adj[v])
            {
                if (!visited[adjV])
                {
                    visited[adjV] = true;
                    s.Push(adjV);
                }
            }
        }

    }

    public static void dfsRecursive(Graph G, int startVert)
    {
        bool[] visited = new bool[G.Size];
        dfsAux(G, startVert, visited);
    }

    private static void dfsAux(Graph G, int v, bool[] visited)
    {
        visited[v] = true;
        foreach(int adjV in G.adj[v])
        {
            if (!visited[adjV])
            {
                // foo(); // Something can happen before the visit.
                dfsAux(G, adjV, visited); // Visiting.
                // bar(); // Something cal happen after the visit.
            }
        }
    }
}

File: TypeScript/Algorithms/DFS.ts
// Simplistic representation of a Graph:
class DFSGraph
{
    public adj: Array<number>[];
    public size: number;
}

class DFS
{
    public dfs(G: DFSGraph, startVert: number)
    {
        let visited: boolean[] = Array<boolean>();
        // Pre-populate array:
        for(let i = 0; i < G.size; i++)
        {
            visited.push(false);
        }

        let s: number[] = new Array();

        visited[startVert] = true;

        s.push(startVert);

        while(s.length > 0)
        {
            const v = s.pop();
            for(let adjV of G.adj[v])
            {
                if(!visited[adjV])
                {
                    visited[adjV] = true;
                    s.push(adjV);
                }
            }
        }
    }

    public dfsRecursive(G: DFSGraph, startVert: number)
    {
        let visited: boolean[] = Array<boolean>();
        // Pre-populate array:
        for(let i = 0; i < G.size; i++)
        {
            visited.push(false);
        }
        this.dfsAux(G, startVert, visited);
    }

    private dfsAux(G: DFSGraph, v: number, visited: boolean[])
    {
        visited[v] = true;
        for(let adjV of G.adj[v])
        {
            if(!visited[adjV])
            {
                // this.foo(); // Something can happen before the visit.
                this.dfsAux(G, adjV, visited);
                // this.bar(); // Something can happen after the visit.
            }
        }
    }
}